### PR TITLE
Tickets/dm 38145

### PR DESCRIPTION
**Pull request by wvreeven**

### CODE REVIEW

File: python/lsst/ts/mtdome/mock_llc/amcs.py [hunk 1]
```diff
@@ -48,6 +53,61 @@
...
+    ----------
+    start_position : `float`
+        The start position.
+    end_position : `float`
+        The end position.
+
+    Returns
+    -------
+    distance: `float`
+        The smallest distance between the initial and target positions.
```
Comment (teweitsai) [inline python/lsst/ts/mtdome/mock_llc/amcs.py:77]: For the physical amounts, please help to add the units. This comment applies to all the code change in this PR.

File: python/lsst/ts/mtdome/mock_llc/amcs.py [hunk 2]
```diff
@@ -48,6 +53,61 @@
...
+
+    Notes
+    -----
+    If "_start_position" and "_end_position" are not in radians, this
```
Comment (teweitsai) [inline python/lsst/ts/mtdome/mock_llc/amcs.py:81]: "_start_position" and "_end_position" --> "start_position" and "end_position"

File: python/lsst/ts/mtdome/mock_llc/amcs.py [hunk 3]
```diff
@@ -48,6 +53,61 @@
...
+    start_position : `float`
+        The start position.
+    end_position : `float`
+        The end position.
+    max_speed : `float`
+        The maximum speed.
+
+    Returns
+    -------
+    duration: `float`
```
Comment (teweitsai) [inline python/lsst/ts/mtdome/mock_llc/amcs.py:105]: duration: `float` -->  duration : `float`

File: python/lsst/ts/mtdome/mock_llc/amcs.py [hunk 4]
```diff
@@ -48,6 +53,61 @@
...
+        The duration of the move, or zero in case of a crawl.
+    """
+    duration = math.fabs(get_distance(start_position, end_position)) / max_speed
+    return duration
```
Comment (teweitsai) [inline python/lsst/ts/mtdome/mock_llc/amcs.py:109]: You can do:

```
return math.fabs(get_distance(start_position, end_position)) / max_speed
```

File: python/lsst/ts/mtdome/mock_llc/amcs.py [hunk 5]
```diff
@@ -98,6 +156,170 @@ def __init__(self, start_tai: float) -> None:
...
+                | InternalMotionState.STATIONARY.name
+            ):
+                await self._handle_motion(current_tai)
+            case MotionState.STOPPED.name:
+                await self._handle_stopped(current_tai)
+            case _:
+                await self._warn_invalid_state()
+
+    async def _warn_invalid_state(self) -> None:
+        self.log.warning(f"Not handling invalid target state {self.target_state}")
```
Comment (teweitsai) [inline python/lsst/ts/mtdome/mock_llc/amcs.py:189]: Function doc string? There are many functions in this PR do not have the doc string. Please consider to add them.

File: python/lsst/ts/mtdome/mock_llc/amcs.py [hunk 6]
```diff
@@ -350,10 +588,19 @@ async def exit_fault(self, start_tai: float) -> float:
         `float`
             The expected duration of the command [s].
         """
-        duration = self.azimuth_motion.exit_fault(start_tai)
-        self.end_tai = start_tai + duration
+        if True in self.drives_in_error_state:
```
Comment (teweitsai) [inline python/lsst/ts/mtdome/mock_llc/amcs.py:591]:
```
if self.drives_in_error_state:
```

File: python/lsst/ts/mtdome/mock_llc/amcs.py [hunk 7]
```diff
@@ -379,9 +626,11 @@ async def reset_drives_az(self, start_tai: float, reset: list[int]) -> float:
         Degraded Mode since the drives don't reset themselves.
         The number of values in the reset parameter is not validated.
         """
-        duration = self.azimuth_motion.reset_drives(start_tai, reset)
-        self.end_tai = start_tai + duration
-        return duration
+        for motor_id, val in enumerate(reset):
+            if val == 1:
+                self.drives_in_error_state[motor_id] = False
```
Comment (teweitsai) [inline python/lsst/ts/mtdome/mock_llc/amcs.py:631]: Is the `reset` an iterable object such as a `list`? This variable name is a little weird to me if you use the `enumerate()` here.

File: python/lsst/ts/mtdome/mock_llc/apscs.py [hunk 8]
```diff
@@ -19,35 +19,71 @@
...
+    start_position : `float`
+        The start position.
+    end_position : `float`
+        The end position.
+    max_speed : `float`
+        The maximum speed.
+
+    Returns
+    -------
+    duration: `float`
```
Comment (teweitsai) [inline python/lsst/ts/mtdome/mock_llc/apscs.py:80]: duration: `float` --> duration : `float`

File: python/lsst/ts/mtdome/mock_llc/apscs.py [hunk 9]
```diff
@@ -283,13 +569,23 @@ async def exit_fault(self, start_tai: float) -> float:
         `float`
             The expected duration of the command [s].
         """
-        durations = []
-        for i in range(NUM_SHUTTERS):
-            durations.append(self.shutter_motion[i].exit_fault(start_tai))
-        duration = max(durations)
-        self.end_tai = start_tai + duration
+        for shutter_id in range(NUM_SHUTTERS):
+            if True in self.drives_in_error_state[shutter_id]:
```
Comment (teweitsai) [inline python/lsst/ts/mtdome/mock_llc/apscs.py:573]: You can use `any()`:

https://www.w3schools.com/python/ref_func_any.asp

This applies to other similar code.

File: python/lsst/ts/mtdome/mock_llc/lwscs.py [hunk 10]
```diff
@@ -25,19 +25,72 @@
...
+    Parameters
+    ----------
+    start_position : `float`
+        The start position.
+    end_position : `float`
+        The end position.
+
+    Returns
+    -------
+    distance: `float`
```
Comment (teweitsai) [inline python/lsst/ts/mtdome/mock_llc/lwscs.py:58]: distance: `float` --> distance : `float`

File: python/lsst/ts/mtdome/mock_llc/lwscs.py [hunk 11]
```diff
@@ -25,19 +25,72 @@
...
+        The smallest distance between the initial and target positions.
+
+    Notes
+    -----
+    If "_start_position" and "_end_position" are not in radians, this
```
Comment (teweitsai) [inline python/lsst/ts/mtdome/mock_llc/lwscs.py:63]: "_start_position" and "_end_position" --> "start_position" and "end_position"

**Review by teweitsai — APPROVED** [review]
Looks good. Some general comments:

1. For the physical amount, please add the unit.
2. Some functions have no doc strings. I am not sure it is because they were the overridden functions or not.
3. The state machines look pretty complex. Is there any document for this? Maybe you could draw them or reference the document in the README file or others.

File: python/lsst/ts/mtdome/mock_llc/amcs.py (hunk 6 above)
Comment (wvreeven) [reply python/lsst/ts/mtdome/mock_llc/amcs.py:591]: That will not work. `self.drives_in_error_state` is an array so the code needs to test if at least one of the elements of the array is True.

File: python/lsst/ts/mtdome/mock_llc/amcs.py (hunk 7 above)
Comment (wvreeven) [reply python/lsst/ts/mtdome/mock_llc/amcs.py:631]: Yes, `reset` is a list. I agree that the name is weird. I'll think of a better one.

File: python/lsst/ts/mtdome/mock_llc/amcs.py (hunk 6 above)
Comment (teweitsai) [reply python/lsst/ts/mtdome/mock_llc/amcs.py]: You can try `any()`. If it is a numpy array, there is an `any()` function as well.

**Pull request merged**